File: tools/check-links.py
#!/usr/bin/env python3
"""ALTER — on-demand link checker (Python port, zero-install).

HEAD-requests every product `url` and hotlinked `img` in data.js (WARDROBE_DATA + SHOP),
follows redirects, and flags anything unreachable. Owned items don't lose ownership when a
link dies — this just tells you which reference links went stale (so the "View" button can be
greyed) and which hotlinked images still need localizing.

Runs with the python3 that ships on macOS — no pip, no dependencies.

    python3 tools/check-links.py                 # check everything
    python3 tools/check-links.py --only=img      # just the hotlinked images
    python3 tools/check-links.py --only=url      # just the product pages
    python3 tools/check-links.py --concurrency=6 --timeout=15000
    python3 tools/check-links.py --dry-run       # plumbing test, no network

Writes tools/link-report.json (full detail) + tools/link-report.md (readable summary).
"""

import json
import os
import re
import ssl
import sys
import urllib.request
from concurrent.futures import ThreadPoolExecutor

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.dirname(HERE)

# ---- args ----------------------------------------------------------------
args = {}
for a in sys.argv[1:]:
    m = re.match(r"^--([^=]+)(?:=(.*))?$", a)
    if m:
        args[m.group(1)] = True if m.group(2) is None else m.group(2)

ONLY = args.get("only", "all")                       # all | url | img
CONCURRENCY = max(1, int(args.get("concurrency", 8)))
TIMEOUT = max(1.0, int(args.get("timeout", 10000)) / 1000.0)   # ms in, seconds out
DRY_RUN = bool(args.get("dry-run"))

HTTP_RE = re.compile(r"^https?://", re.I)


def is_http(s):
    return isinstance(s, str) and bool(HTTP_RE.match(s))


# ---- extract items from data.js ------------------------------------------
# Only WARDROBE_DATA + SHOP items carry url/img. Objects can span multiple lines (the SHOP
# items do), so we can't scan line-by-line — we pull each array's body, split it into top-level
# object literals with a string/brace-aware scanner, then regex the fields off each object.
FIELD = lambda key, text: (re.search(r'\b' + key + r':"((?:[^"\\]|\\.)*)"', text) or [None, None])[1]


def _extract_array_body(src, name):
    """Return the inner text between the `[` and matching `]` of `const NAME = [ ... ]`."""
    m = re.search(r'\b' + re.escape(name) + r'\s*=\s*\[', src)
    if not m:
        return ""
    i = m.end() - 1               # index of the opening '['
    depth, instr, esc, start = 0, False, False, i + 1
    while i < len(src):
        ch = src[i]
        if instr:
            if esc: esc = False
            elif ch == "\\": esc = True
            elif ch == '"': instr = False
        elif ch == '"':
            instr = True
        elif ch == "[":
            depth += 1
        elif ch == "]":
            depth -= 1
            if depth == 0:
                return src[start:i]
        i += 1
    return ""


def _split_objects(body):
    """Split an array body into its top-level {...} object literals (string/brace-aware)."""
    objs, depth, instr, esc, start = [], 0, False, False, None
    for i, ch in enumerate(body):
        if instr:
            if esc: esc = False
            elif ch == "\\": esc = True
            elif ch == '"': instr = False
            continue
        if ch == '"':
            instr = True
        elif ch == "{":
            if depth == 0:
                start = i
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0 and start is not None:
                objs.append(body[start:i + 1])
                start = None
    return objs


def _extract_object_body(src, name):
    """Inner text between `{` and matching `}` of `const NAME = { ... }`."""
    m = re.search(r'\b' + re.escape(name) + r'\s*=\s*\{', src)
    if not m:
        return ""
    i, depth, instr, esc, start = m.end() - 1, 0, False, False, m.end()
    while i < len(src):
        ch = src[i]
        if instr:
            if esc: esc = False
            elif ch == "\\": esc = True
            elif ch == '"': instr = False
        elif ch == '"':
            instr = True
        elif ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                return src[start:i]
        i += 1
    return ""


def load_targets():
    src = open(os.path.join(REPO, "data.js"), encoding="utf-8").read()
    items = []
    for arr_name in ("WARDROBE_DATA", "SHOP"):
        for obj in _split_objects(_extract_array_body(src, arr_name)):
            item_id = FIELD("id", obj)
            if not item_id:
                continue
            images = []
            arr = re.search(r'\bimages:\[([^\]]*)\]', obj)
            if arr:
                images = re.findall(r'"((?:[^"\\]|\\.)*)"', arr.group(1))
            items.append({
                "id": item_id,
                "name": FIELD("name", obj) or item_id,
                "brand": FIELD("brand", obj) or "",
                "url": FIELD("url", obj),
                "img": FIELD("img", obj),
                "images": images,
            })
    # IMAGES { "name":"url" } map — the primary per-item image source (pieceImg reads it first).
    img_map = re.findall(r'"((?:[^"\\]|\\.)*)"\s*:\s*"((?:[^"\\]|\\.)*)"',
                         _extract_object_body(src, "IMAGES"))
    return items, [{"name": k, "url": v} for k, v in img_map]


def collect(items, images_map):
    targets = {}   # url -> {url, kind, items:[...]}
    local_images = []

    def add(url, kind, ref):
        targets.setdefault(url, {"url": url, "kind": kind, "items": []})["items"].append(ref)

    for it in items:
        ref = {"id": it["id"], "name": it["name"], "brand": it["brand"]}
        if ONLY != "img" and is_http(it["url"]):
            add(it["url"], "url", ref)
        if ONLY != "url":
            for img in ([it["img"]] if it["img"] else []) + it["images"]:
                if is_http(img):
                    add(img, "img", ref)
                elif img:
                    local_images.append({**ref, "img": img})
    if ONLY != "url":
        for e in images_map:
            ref = {"id": "IMAGES", "name": e["name"], "brand": ""}
            if is_http(e["url"]):
                add(e["url"], "img", ref)
            elif e["url"]:
                local_images.append({**ref, "img": e["url"]})
    return targets, local_images


# ---- probe ---------------------------------------------------------------
CTX = ssl.create_default_context()
UA = "Mozilla/5.0 (compatible; AlterLinkCheck/1.0)"


def _request(url, method):
    req = urllib.request.Request(url, method=method, headers={"User-Agent": UA})
    with urllib.request.urlopen(req, timeout=TIMEOUT, context=CTX) as r:
        return r.status, r.geturl()


def probe(url):
    try:
        try:
            status, final = _request(url, "HEAD")
        except urllib.error.HTTPError as e:
            # some servers reject HEAD — retry GET before trusting the failure
            if e.code in (403, 405, 501):
                status, final = _request(url, "GET")
            else:
                status, final = e.code, url
        redirected = final and final.rstrip("/") != url.rstrip("/")
        state = "ok"
        if status >= 400:
            state = "dead"
        elif status >= 300:
            state = "redirect"
        return {"state": state, "status": status, "finalUrl": final, "redirected": redirected}
    except urllib.error.HTTPError as e:
        return {"state": "dead", "status": e.code, "reason": e.reason}
    except Exception as e:  # URLError / timeout / SSL / DNS
        reason = getattr(e, "reason", None) or str(e)
        return {"state": "unreachable", "status": 0, "reason": str(reason)}


# ---- report --------------------------------------------------------------
def item_list(t):
    return "; ".join(f"{i['name']} ({i['brand']}, {i['id']})" for i in t["items"])


def to_markdown(rep):
    s = rep["summary"]
    L = ["# Link report", "", f"_{s['generatedAt']} · scope: {s['scope']}_", "",
         "| | count |", "|---|---:|",
         f"| Items scanned | {s['items']} |",
         f"| URLs checked | {s['checked']} |",
         f"| OK | {s['ok']} |",
         f"| Redirect | {s['redirect']} |",
         f"| Dead (4xx/5xx) | {s['dead']} |",
         f"| Unreachable | {s['unreachable']} |",
         f"| Hotlinked images (need localizing) | {s['hotlinkedImages']} |",
         f"| Already-local images | {s['localImages']} |", ""]

    dead = rep["dead"]
    if dead:
        L += ["## Dead / unreachable — needs attention", ""]
        for t in dead:
            reason = f" · {t.get('reason')}" if t.get("reason") else ""
            L.append(f"- **[{t['kind']}] {t.get('status') or t['state']}{reason}** — {t['url']}")
            L.append(f"  - used by: {item_list(t)}")
        L.append("")
    else:
        L += ["## Dead / unreachable", "", "None.", ""]

    if rep["redirects"]:
        L += ["## Redirects (still resolve, but the stored URL moved)", ""]
        for t in rep["redirects"]:
            L.append(f"- {t['url']} -> {t.get('finalUrl')} — {item_list(t)}")
        L.append("")

    if rep["httpImgs"]:
        L += [f"## Hotlinked images still to localize ({len(rep['httpImgs'])})", "",
              "These `img` values point at remote CDNs and will break when the listing rotates.", ""]
        for t in rep["httpImgs"]:
            st = f" {t.get('status')}" if t.get("status") else ""
            L.append(f"- [{t['state']}{st}] {t['url']} — {item_list(t)}")
        L.append("")

    if rep["localImages"]:
        L += [f"## Already local (safe) — {len(rep['localImages'])}", ""]
        for i in rep["localImages"]:
            L.append(f"- {i['img']} — {i['name']} ({i['brand']}, {i['id']})")
        L.append("")
    return "\n".join(L)


# ---- main ----------------------------------------------------------------
def main():
    from datetime import datetime, timezone
    items, images_map = load_targets()
    targets, local_images = collect(items, images_map)
    print(f"Loaded {len(items)} items + {len(images_map)} IMAGES entries -> {len(targets)} unique link(s) "
          f"to check ({len(local_images)} already local).")

    values = list(targets.values())
    if DRY_RUN or not values:
        for t in values:
            t.update(state="skipped", status=0)
        if DRY_RUN:
            print("(--dry-run: no network calls made)")
    else:
        done = [0]
        total = len(values)
        print("Checking", end="", flush=True)

        def work(t):
            t.update(probe(t["url"]))
            done[0] += 1
            if done[0] % 5 == 0 or done[0] == total:
                print(f" {done[0]}/{total}", end="", flush=True)

        with ThreadPoolExecutor(max_workers=CONCURRENCY) as pool:
            list(pool.map(work, values))
        print()

    by = lambda st: [t for t in values if t["state"] == st]
    dead = sorted(by("dead") + by("unreachable"), key=lambda t: t["kind"])
    http_imgs = [t for t in values if t["kind"] == "img"]
    summary = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "scope": ONLY, "items": len(items), "checked": len(values),
        "ok": len(by("ok")), "redirect": len(by("redirect")),
        "dead": len(by("dead")), "unreachable": len(by("unreachable")),
        "hotlinkedImages": len(http_imgs), "localImages": len(local_images),
    }
    rep = {"summary": summary, "list": values, "dead": dead,
           "redirects": by("redirect"), "httpImgs": http_imgs, "localImages": local_images}

    open(os.path.join(HERE, "link-report.json"), "w").write(json.dumps(rep, indent=2))
    open(os.path.join(HERE, "link-report.md"), "w").write(to_markdown(rep))

    print(f"\nDone. ok:{summary['ok']}  redirect:{summary['redirect']}  "
          f"dead:{summary['dead']}  unreachable:{summary['unreachable']}  "
          f"hotlinked-img:{summary['hotlinkedImages']}\n"
          f"Report -> tools/link-report.md (+ .json)")


if __name__ == "__main__":
    main()

File: tools/check-links.mjs
#!/usr/bin/env node
/* ALTER — on-demand link checker.
 *
 * HEAD-requests every product `url` and hotlinked `img` in data.js (WARDROBE_DATA + SHOP),
 * follows redirects, and flags anything that isn't reachable. Owned items don't lose their
 * ownership when a link dies — this just tells you which reference links have gone stale
 * (so the "View" button can be greyed) and which hotlinked images still need localizing.
 *
 * RUN LOCALLY (needs open network egress — will NOT work inside the Cowork sandbox, whose
 * proxy blocks arbitrary domains):
 *
 *     node tools/check-links.mjs                # check everything
 *     node tools/check-links.mjs --only=img     # just the hotlinked images
 *     node tools/check-links.mjs --only=url     # just the product pages
 *     node tools/check-links.mjs --concurrency=6 --timeout=12000
 *     node tools/check-links.mjs --json         # machine-readable dump to stdout too
 *
 * Writes tools/link-report.json (full detail) + tools/link-report.md (readable summary).
 * Requires Node 18+ (global fetch). No dependencies.
 */

import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import vm from "node:vm";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO = join(__dirname, "..");

// ---- args ---------------------------------------------------------------
const args = Object.fromEntries(
  process.argv.slice(2).map((a) => {
    const m = a.match(/^--([^=]+)(?:=(.*))?$/);
    return m ? [m[1], m[2] === undefined ? true : m[2]] : [a, true];
  })
);
const ONLY = args.only || "all";               // all | url | img
const CONCURRENCY = Math.max(1, parseInt(args.concurrency || "8", 10));
const TIMEOUT = Math.max(1000, parseInt(args.timeout || "10000", 10));
const EMIT_JSON = !!args.json;

// ---- load the data arrays out of data.js via a sandboxed vm -------------
// data.js declares top-level `const WARDROBE_DATA = [...]` etc. and defines functions that
// reference browser globals — but those bodies never execute at load, so stub globals + a
// capture line are all we need. (const is script-scoped, so we append an explicit capture.)
function loadData() {
  const src = readFileSync(join(REPO, "data.js"), "utf8");
  const captured = {};
  const sandbox = {
    __OUT: captured,
    console: { log() {}, warn() {}, error() {} },
    document: {}, window: {}, navigator: {}, location: {},
    localStorage: { getItem: () => null, setItem() {}, removeItem() {} },
    setTimeout() {}, clearTimeout() {}, fetch: () => Promise.reject(new Error("no net at load")),
  };
  sandbox.window = sandbox;
  vm.createContext(sandbox);
  const capture = `
    ;try { __OUT.WARDROBE_DATA = typeof WARDROBE_DATA !== "undefined" ? WARDROBE_DATA : null; } catch(e) {}
    ;try { __OUT.SHOP = typeof SHOP !== "undefined" ? SHOP : null; } catch(e) {}
  `;
  try {
    new vm.Script(src + capture, { filename: "data.js" }).runInContext(sandbox, { timeout: 5000 });
  } catch (e) {
    console.error("Could not evaluate data.js:", e.message);
    process.exit(1);
  }
  return {
    wardrobe: Array.isArray(captured.WARDROBE_DATA) ? captured.WARDROBE_DATA : [],
    shop: Array.isArray(captured.SHOP) ? captured.SHOP : [],
  };
}

const isHttp = (s) => typeof s === "string" && /^https?:\/\//i.test(s);

// Flatten every image reference an item might carry now or later.
function itemImages(it) {
  const imgs = [];
  if (it.img) imgs.push(it.img);
  if (Array.isArray(it.images)) imgs.push(...it.images);
  return imgs;
}

// ---- collect the targets ------------------------------------------------
function collectTargets({ wardrobe, shop }) {
  const items = [
    ...wardrobe.map((it) => ({ ...it, _source: "wardrobe" })),
    ...shop.map((it) => ({ ...it, _source: "shop" })),
  ];
  // unique URL -> { url, kind, items:[{id,name,brand,source}] }
  const targets = new Map();
  const localImages = []; // already localized (filename, not http) — reported, not fetched

  for (const it of items) {
    const ref = { id: it.id, name: it.name, brand: it.brand, source: it._source };
    if (ONLY !== "img" && isHttp(it.url)) {
      addTarget(targets, it.url, "url", ref);
    }
    if (ONLY !== "url") {
      for (const img of itemImages(it)) {
        if (isHttp(img)) addTarget(targets, img, "img", ref);
        else if (img) localImages.push({ ...ref, img });
      }
    }
  }
  return { targets, localImages, itemCount: items.length };
}

function addTarget(map, url, kind, ref) {
  if (!map.has(url)) map.set(url, { url, kind, items: [] });
  map.get(url).items.push(ref);
}

// ---- fetch with timeout + HEAD→GET fallback -----------------------------
async function probe(url) {
  const attempt = async (method) => {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), TIMEOUT);
    try {
      const r = await fetch(url, {
        method,
        redirect: "follow",
        signal: ctrl.signal,
        headers: { "user-agent": "Mozilla/5.0 (compatible; AlterLinkCheck/1.0)" },
      });
      return { status: r.status, finalUrl: r.url, ok: r.ok };
    } finally {
      clearTimeout(timer);
    }
  };
  try {
    let res = await attempt("HEAD");
    // Some servers reject/΄405 HEAD — retry once with GET before trusting the failure.
    if (res.status === 405 || res.status === 501 || res.status === 403) {
      try { res = await attempt("GET"); } catch { /* keep HEAD result */ }
    }
    const redirected = res.finalUrl && res.finalUrl.replace(/\/$/, "") !== url.replace(/\/$/, "");
    let state = "ok";
    if (res.status >= 400) state = "dead";
    else if (res.status >= 300) state = "redirect";
    return { state, status: res.status, finalUrl: res.finalUrl, redirected };
  } catch (e) {
    const reason = e.name === "AbortError" ? "timeout" : (e.cause && (e.cause.code || e.cause.message)) || e.message;
    return { state: "unreachable", status: 0, reason: String(reason) };
  }
}

async function runPool(targets, onDone) {
  const list = [...targets.values()];
  let i = 0;
  const workers = Array.from({ length: CONCURRENCY }, async () => {
    while (i < list.length) {
      const t = list[i++];
      const result = await probe(t.url);
      Object.assign(t, result);
      onDone(t, i, list.length);
    }
  });
  await Promise.all(workers);
  return list;
}

// ---- reporting ----------------------------------------------------------
function buildReport(list, localImages, meta) {
  const by = (state) => list.filter((t) => t.state === state);
  const dead = [...by("dead"), ...by("unreachable")].sort((a, b) => a.kind.localeCompare(b.kind));
  const redirects = by("redirect");
  const httpImgs = list.filter((t) => t.kind === "img");
  const summary = {
    generatedAt: new Date().toISOString(),
    scope: ONLY,
    items: meta.itemCount,
    checked: list.length,
    ok: by("ok").length,
    redirect: redirects.length,
    dead: by("dead").length,
    unreachable: by("unreachable").length,
    hotlinkedImages: httpImgs.length,
    localImages: localImages.length,
  };
  return { summary, list, dead, redirects, httpImgs, localImages };
}

function toMarkdown(rep) {
  const { summary: s, dead, redirects, httpImgs, localImages } = rep;
  const L = [];
  L.push(`# Link report`);
  L.push(``);
  L.push(`_${s.generatedAt} · scope: ${s.scope}_`);
  L.push(``);
  L.push(`| | count |`);
  L.push(`|---|---:|`);
  L.push(`| Items scanned | ${s.items} |`);
  L.push(`| URLs checked | ${s.checked} |`);
  L.push(`| ✅ OK | ${s.ok} |`);
  L.push(`| ↪️ Redirect | ${s.redirect} |`);
  L.push(`| ❌ Dead (4xx/5xx) | ${s.dead} |`);
  L.push(`| ⛔ Unreachable | ${s.unreachable} |`);
  L.push(`| 🖼️ Hotlinked images (need localizing) | ${s.hotlinkedImages} |`);
  L.push(`| 📁 Already-local images | ${s.localImages} |`);
  L.push(``);

  const itemList = (t) => t.items.map((i) => `${i.name} (${i.brand}, ${i.id}, ${i.source})`).join("; ");

  if (dead.length) {
    L.push(`## ❌ Dead / unreachable — needs attention`);
    L.push(``);
    for (const t of dead) {
      L.push(`- **[${t.kind}] ${t.status || t.state}${t.reason ? " · " + t.reason : ""}** — ${t.url}`);
      L.push(`  - used by: ${itemList(t)}`);
    }
    L.push(``);
  } else {
    L.push(`## ❌ Dead / unreachable\n\nNone. 🎉\n`);
  }

  if (redirects.length) {
    L.push(`## ↪️ Redirects (still resolve, but the stored URL moved)`);
    L.push(``);
    for (const t of redirects) {
      L.push(`- ${t.url} → ${t.finalUrl} — ${itemList(t)}`);
    }
    L.push(``);
  }

  if (httpImgs.length) {
    L.push(`## 🖼️ Hotlinked images still to localize (${httpImgs.length})`);
    L.push(``);
    L.push(`These \`img\` values point at remote CDNs and will break when the listing rotates.`);
    L.push(``);
    for (const t of httpImgs) {
      L.push(`- [${t.state}${t.status ? " " + t.status : ""}] ${t.url} — ${itemList(t)}`);
    }
    L.push(``);
  }

  if (localImages.length) {
    L.push(`## 📁 Already local (safe) — ${localImages.length}`);
    L.push(``);
    for (const i of localImages) L.push(`- ${i.img} — ${i.name} (${i.brand}, ${i.id})`);
    L.push(``);
  }
  return L.join("\n");
}

// ---- main ---------------------------------------------------------------
const NET = !(args["dry-run"] || args.dryRun); // --dry-run skips fetching (verify plumbing without egress)
const data = loadData();
const { targets, localImages, itemCount } = collectTargets(data);

console.log(
  `Loaded ${data.wardrobe.length} wardrobe + ${data.shop.length} shop items → ` +
  `${targets.size} unique link${targets.size === 1 ? "" : "s"} to check ` +
  `(${localImages.length} already local).`
);

let list;
if (NET && targets.size) {
  process.stdout.write("Checking");
  list = await runPool(targets, (_t, done, total) => {
    if (done % 5 === 0 || done === total) process.stdout.write(` ${done}/${total}`);
  });
  process.stdout.write("\n");
} else {
  // dry-run / nothing to check: mark every target "skipped" so the report still renders.
  list = [...targets.values()].map((t) => ({ ...t, state: NET ? "ok" : "skipped", status: 0 }));
  if (!NET) console.log("(--dry-run: no network calls made)");
}

const rep = buildReport(list, localImages, { itemCount });
writeFileSync(join(__dirname, "link-report.json"), JSON.stringify(rep, null, 2));
writeFileSync(join(__dirname, "link-report.md"), toMarkdown(rep));

console.log(
  `\nDone. ok:${rep.summary.ok}  redirect:${rep.summary.redirect}  ` +
  `dead:${rep.summary.dead}  unreachable:${rep.summary.unreachable}  ` +
  `hotlinked-img:${rep.summary.hotlinkedImages}\n` +
  `Report → tools/link-report.md (+ .json)`
);
if (EMIT_JSON) console.log(JSON.stringify(rep.summary, null, 2));
